federation/pkg/federation-controller/util/federated_informer.go

```diff
@@ -0,0 +1,408 @@
+	"k8s.io/kubernetes/pkg/controller/framework"
+	pkg_runtime "k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+const (
+	clusterSyncPeriod = 10 * time.Minute
+	UserAgentName     = "federation-service-controller"
+)
+
+// FederatedReadOnlyStore is an overlay over multiple stores created in federated clusters.
+type FederatedReadOnlyStore interface {
+
+	// Returns all items in the store.
+	List() []interface{}
+
+	// GetByKey returns the item stored under the given key in the specified cluster (if exist).
+	GetByKey(clusterName string, key string) (item interface{}, exists bool)
+
+	// Returns the items stored under the given key in all clusters.
+	GetFromAllClusters(key string) []interface{}
+
+	// Checks whether stores for all clusters form the lists (and only these) are there and
+	// are synced. This is only a basic check whether the data inside of the store is usable.
+	// It is not a full synchronization/locking mechanism it only tries to ensure that out-of-sync
+	// issues occur less often.	All users of the interface should assume
+	// that there may be significant delays in content updates of all kinds and write their
+	// code that it doesn't break if something is slightly out-of-sync.
+	Synced(clusters []string) bool
+}
+
+// An interface to access federation members and clients.
+type FederationView interface {
+	// GetClientsetForCluster returns a clientset for the cluster, if present.
+	GetClientsetForCluster(clusterName string) (federation_release_1_4.Interface, error)
+
+	// GetClusers returns all clusters for which the sub-informers are run.
+	GetClusters() []*federation_api.Cluster
+
+	// GetReadyCluster returns the cluster with the given name, if found.
+	GetCluster(name string) (*federation_api.Cluster, bool)
+
+	// Synced returns true if the view is synced (for the first time)
+	ClusterSynced() bool
+}
+
+// A tyoe of structure that combines an infomer running agains federated api server and listening for cluster updates
+// with multiple Kubernetes API informers (called target infomers) running against federation members. Whenever a new
+// cluster is added to the federation an infomer is created for it using TargetInformerFactory. Infomers are stoped
+// when a cluster is either put offline of deleted. It is assumed that some controller keeps an eye on the cluster list
+// and thus the clusters in ETCD are up to date.
+type FederatedInfomer interface {
+	FederationView
+
+	// Returns a store created over all stores from target informers.
+	GetTargetStore() FederatedReadOnlyStore
+
+	// Starts all the processes.
+	Start()
+
+	// Stops all the processes inside the infomer.
+	Stop()
+}
+
+// The function that should be used to create an infomer on the target object. Store should use
+// framework.DeletionHandlingMetaNamespaceKeyFunc as a keying function.
+type TargetInformerFactory func(*federation_api.Cluster, federation_release_1_4.Interface) (cache.Store, *framework.Controller)
```

Comment:
Can we use "framework.ControllerInterface" instead of "*framework.Controller" ?
[We are trying to get rid of the latter, so if it's possible, please change, otherwise please comment why we need the latter.]